src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1110 @@
+        //
+        // Be careful when modifying this, calculate_layout relies on the
+        // overflow check here.
+        cap.checked_mul(8)? / 7
+    };
+
+    // Any overflows will have been caught by the checked_mul.
+    Some(adjusted_cap.next_power_of_two())
+}
+
+/// Returns the maximum effective capacity for the given bucket mask, taking
+/// the maximum load factor into account.
+#[inline]
+fn bucket_mask_to_capacity(bucket_mask: usize) -> usize {
+    if bucket_mask < 8 {
+        bucket_mask
+    } else {
+        ((bucket_mask + 1) / 8) * 7
+    }
+}
+
+// Returns a Layout which describes the allocation required for a hash table,
+// and the offset of the buckets in the allocation.
+///
+/// Returns `None` if an overflow occurs.
+#[inline]
+fn calculate_layout<T>(buckets: usize) -> Option<(Layout, usize)> {
+    debug_assert!(buckets.is_power_of_two());
+
+    // Array of buckets
+    let data = Layout::array::<T>(buckets).ok()?;
+
+    // Array of control bytes. This must be aligned to the group size.
+    //
+    // We add `Group::WIDTH` control bytes at the end of the array which
+    // replicate the bytes at the start of the array and thus avoids the need to
+    // perform bounds-checking while probing.
+    //
+    // There is no possible overflow here since buckets is a power of two and
+    // Group::WIDTH is a small number.
+    let ctrl = unsafe { Layout::from_size_align_unchecked(buckets + Group::WIDTH, Group::WIDTH) };
+
+    ctrl.extend(data).ok()
+}
+
+/// A reference to a hash table bucket containing a `T`.
+pub struct Bucket<T> {
+    ptr: NonNull<T>,
+}
+
+impl<T> Clone for Bucket<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Bucket { ptr: self.ptr }
+    }
+}
+
+impl<T> Bucket<T> {
+    #[inline]
+    unsafe fn from_ptr(ptr: *const T) -> Self {
+        Bucket {
+            ptr: NonNull::new_unchecked(ptr as *mut T),
+        }
+    }
+    #[inline]
+    pub unsafe fn drop(&self) {
+        self.ptr.as_ptr().drop_in_place();
+    }
+    #[inline]
+    pub unsafe fn read(&self) -> T {
+        self.ptr.as_ptr().read()
+    }
+    #[inline]
+    pub unsafe fn write(&self, val: T) {
+        self.ptr.as_ptr().write(val);
+    }
+    #[inline]
+    pub unsafe fn as_ref<'a>(&self) -> &'a T {
+        &*self.ptr.as_ptr()
+    }
+    #[inline]
+    pub unsafe fn as_mut<'a>(&self) -> &'a mut T {
+        &mut *self.ptr.as_ptr()
+    }
+}
+
+/// A raw hash table with an unsafe API.
+pub struct RawTable<T> {
+    ctrl: NonNull<u8>,
+    bucket_mask: usize,
+    data: NonNull<T>,
+    items: usize,
+    growth_left: usize,
+}
+
+impl<T> RawTable<T> {
+    /// Creates a new empty hash table without allocating any memory.
+    ///
+    /// In effect this returns a table with exactly 1 bucket. However we can
+    /// leave the data pointer dangling since that bucket is never written to
+    /// due to our load factor forcing us to always have at least 1 free bucket.
```

Comment:
Because of load factor, a table with 1 bucket has a capacity of 0. This means that the table will be resized before insertion. And remove will never reach this since the table is always empty so there is nothing to remove.